src/nnet3/discriminative-training.cc

```diff
@@ -0,0 +1,602 @@
+// limitations under the License.
+
+#include "nnet3/discriminative-training.h"
+#include "lat/lattice-functions.h"
+#include "cudamatrix/cu-matrix.h"
+
+namespace kaldi {
+namespace discriminative {
+
+DiscriminativeTrainingStats::DiscriminativeTrainingStats() {
+  std::memset(this, 0, sizeof(*this));
+  config.accumulate_gradients = false; 
+  config.accumulate_output = false; 
+  config.num_pdfs = 0; 
+}
+
+DiscriminativeTrainingStats::DiscriminativeTrainingStats(int32 num_pdfs) {
+  config.accumulate_gradients = false; 
+  config.accumulate_output = false; 
+  config.num_pdfs = num_pdfs; 
+  gradients.Resize(num_pdfs); 
+  output.Resize(num_pdfs);
+}
+
+// Constructor from config structure
+DiscriminativeTrainingStats::DiscriminativeTrainingStats(
+    DiscriminativeTrainingStatsOptions opts) : config(opts) { 
+  gradients.Resize(opts.num_pdfs); 
+  output.Resize(opts.num_pdfs);
+}
+
+// Reset statistics
+void DiscriminativeTrainingStats::Reset() {
+  gradients.SetZero();
+  output.SetZero();
+
+  tot_t = 0.0;
+  tot_t_weighted = 0.0;
+  tot_objf = 0.0;
+  tot_num_count = 0.0;
+  tot_den_count = 0.0;
+  tot_num_objf = 0.0;
+}
+
+// Sets the config structure
+void DiscriminativeTrainingStats::SetConfig(
+    const DiscriminativeTrainingStatsOptions &opts) {
+  config = opts;
+  gradients.Resize(opts.num_pdfs); 
+  output.Resize(opts.num_pdfs);
+}
+
+// This class is responsible for the forward-backward of the
+// 'supervision' lattices and computation of the objective function
+// and gradients.
+//
+// note: the supervision.weight is ignored by this class, you have to apply
+// it externally.
+class DiscriminativeComputation {
+  typedef Lattice::Arc Arc;
+  typedef Arc::StateId StateId;
+ 
+ public:
+  // Initialize the objcect.  Note: we expect the 'nnet_output' to have the
+  // same number of rows as supervision.num_frames * supervision.num_sequences,
+  // and the same number of columns as tmodel.NumPdfs(); but the
+  // ordering of the rows of 'nnet_output' is not the same as the ordering of
+  // frames in paths in the 'supervision' object (which has all frames of the
+  // 1st sequence first, then the 2nd sequence, and so on).  Instead, the
+  // frames in 'nnet_output' are ordered as: first the first frame of each
+  // sequence, then the second frame of each sequence, and so on. 
+  // This is done to be similar to the setup in 'chain' training 
+  // even though this does not offer any computational advantages here
+  // as in the 'chain' case.
+  DiscriminativeComputation(const DiscriminativeTrainingOptions &opts,
+      const TransitionModel &tmodel,
+      const CuVectorBase<BaseFloat> &log_priors,
+      const DiscriminativeSupervision &supervision,
+      const CuMatrixBase<BaseFloat> &nnet_output,
+      DiscriminativeTrainingStats *stats,
+      BaseFloat *l2_term,
+      CuMatrixBase<BaseFloat> *nnet_output_deriv,
+      CuMatrixBase<BaseFloat> *xent_output_deriv);
```

Comment:
On Tue, Mar 8, 2016 at 1:10 AM, Vimal Manohar <notifications@github.com>
wrote:

> In src/nnet3/discriminative-training.cc
> <https://github.com/kaldi-asr/kaldi/pull/563#discussion_r55320043>:
>
> > +  // frames in paths in the 'supervision' object (which has all frames of the
> > +  // 1st sequence first, then the 2nd sequence, and so on).  Instead, the
> > +  // frames in 'nnet_output' are ordered as: first the first frame of each
> > +  // sequence, then the second frame of each sequence, and so on.
> > +  // This is done to be similar to the setup in 'chain' training
> > +  // even though this does not offer any computational advantages here
> > +  // as in the 'chain' case.
> > +  DiscriminativeComputation(const DiscriminativeTrainingOptions &opts,
> > +      const TransitionModel &tmodel,
> > +      const CuVectorBase<BaseFloat> &log_priors,
> > +      const DiscriminativeSupervision &supervision,
> > +      const CuMatrixBase<BaseFloat> &nnet_output,
> > +      DiscriminativeTrainingStats *stats,
> > +      BaseFloat *l2_term,
> > +      CuMatrixBase<BaseFloat> *nnet_output_deriv,
> > +      CuMatrixBase<BaseFloat> *xent_output_deriv);
>
> I had assumed by "why not use the single output of the nnet, which already
> has a softmax, for both purposes", you meant for me to rename "output-xent"
> as "output" and use that for sequence training.
>

I was talking about the non-chain case.


> This gives poor results (actually only 2% worse WER after tuning acwt and
> post-decode-acwt) than using "output". So it is better to use "output".
> Hence we need xent_output_deriv. We can even use this with CE models that
> don't have "output-xent" by modifying the nnet config prior to sequence
> training. This way we can use the same code for both chain and non-chain
> models.
>
That's an odd thing to do, in my opinion.  And since your experiments seem
to be finding that discriminative training does not even work for the chain
models, it's not clear that the chain models should be dominating the
design of this code.

Dan




> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/563/files#r55320043>.
>